### ISSUE TITLE
MCS - TCB Affinity Modification Bug

### ISSUE THREAD

**Issue by chrisguikema** [seL4/seL4#138]
I've noticed an issue on the MCS branch with changing a threads core by invoking the `seL4_SchedControl_Configure` system call. Essentially, if a thread is running on the new core, the TCB will migrate properly. However, if no thread is active on the new core, the TCB migrates, but a scheduler dump shows the TCB as "running", but its instruction pointer is the `syscall` instruction, and everything stalls out. This behavior is also seen with passive servers.

For example, the camkes-vm can be configured such that the Serial Server provides passive interfaces for `putchar` and `getchar`. With a single VM, this functions as expected. However, with multiple VMs on different cores, the same behavior is encountered. The passive server is shown to be "Running" at the `syscall` instruction. This obviously causes issues, because the calling thread has donated its scheduling context, and can no longer run.

By reverting the changes in this commit, https://github.com/seL4/seL4/commit/ef47fa6e3c842112bbe8fbfc0783af9f78088a4c, I can get the passive servers to run on QEMU, but not on real hardware (Optiplex, i7). 

Do you have any insight on where I can look or what is happening? I can attempt to spend some time and fix the issue.

**Comment by chrisguikema** [comment]
As an update, it appears that the system behaves differently on QEMU v. Hardware even when not functioning.

This is a (slightly modified) dump scheduler state of the system on QEMU:
```
Dumping all tcbs on Core 1!
Name                                              State               IP                  Prio 	Core 	HasSchedContext InReleaseQueue
--------------------------------------------------------------------------------------------------------------------------------------
serial:processed_putchar                running             0x406778            100	1	1		0
vm0:control                                    blocked on reply    0x41b19a       100	1	0		0
```
So at this point on QEMU, the context has been donated, but is stuck.

And on hardware:
```
Dumping all tcbs on Core 0!
--------------------------------------------------------------------------------------------------------------------------------------
serial:processed_putchar                          running             0x406778            100   0       0               0

Dumping all tcbs on Core 1!
Name                                              State               IP                  Prio  Core    HasSchedContext InReleaseQueue
--------------------------------------------------------------------------------------------------------------------------------------
vm0:control                                       blocked on reply    0x41b19a            100   1       1               0
```

So basically, QEMU seems to be behaving as expected. Adding `SMP_COND_STATEMENT(remoteQueueUpdate(to));` after `    SMP_COND_STATEMENT(migrateTCB(to, sc->scCore));` actually gets QEMU fully functional. But we never even reach that point on hardware.

Does anyone have any insight here?

**Comment by chrisguikema** [comment]
(Hopefully) The final update.

As it turns out, the passive servers are still in the process of initializing when each VM is attempting to use them. This means they have a scheduling context when the VMs come calling. The solution worked on QEMU because QEMU's secondary cores take longer to boot up.

I tested with interfaces that don't get used immediately, and passive servers worked and changed cores as expected.

So basically, add `SMP_COND_STATEMENT(remoteQueueUpdate(to));` during scheduling context donation and it fixes the issue.

**Comment by pingerino** [comment]
Hi Chris, 

Nice tracking this down. Servers do need synchronisation when using donation or everything can end up in a weird state. AFAICT the `remoteQueueUpdate` shouldn't be required in `to` in `schedContext_donate` as the calling site of that function should handle notifying the core (specifically in `possibleSwitchTo`).

**Comment by chrisguikema** [comment]
Hey Anna, 

I think in the use case I'm looking at, `remoteQueueUpdate` is required, especially after `migrateTCB` is called (But I could be mistaken). If `to` is running on a different core than the scheduling context, then `to->tcbAffinity` will be updated in `migrateTCB`. At that point, `to->tcbAffinity` would be `y`, while the current core would be `x`. `remoteQueueUpdate` first checks for `tcb->tcbAffinity != getCurrentCPUIndex()`. So if this was called before `migrateTCB`, the inner code block wouldn't be called, which triggers a rescheduling if the core is idle, or the priority of the new thread is higher than the current thread's priority.

**Comment by chrisguikema** [comment]
Is there any update on this issue? It seems like adding a test to seL4test for a process switching cores would shed some light on what is happening

**Comment by xurtis** [comment]
Hi Chris,

Do you think you could write up some pseudo-code for the behavior of the two threads and where you think they both get stuck? This should help us put together a test case we can put into `sel4test` that replicates the behavior.

**Comment by chrisguikema** [comment]
Sure,

Here is some example code where I've seen the issue.

```static void change_core_affinity(int core)
{
    int error = seL4_SchedControl_Configure(simple_get_sched_ctrl(&camkes_simple, core),
                                            camkes_get_tls()->sc_cap, 10000, 10000, 1, 0);
    ZF_LOGF_IF(error, "Failed to configure scheduling context");
}
```

This would cause the system to freeze when changing to a core with nothing running. It should be fairly easy to replicate using the capDL loader, or the rootserver with seL4test. Since they run on core 0, changing to core 1 _should_ trigger the issue. 

And this was the code I used to fix the issue:

```diff --git a/src/object/schedcontext.c b/src/object/schedcontext.c
index e9efdeac..d1b3bcf5 100644
--- a/src/object/schedcontext.c
+++ b/src/object/schedcontext.c
@@ -334,7 +334,14 @@ schedContext_donate(sched_context_t *sc, tcb_t *to)
     sc->scTcb = to;
     to->tcbSchedContext = sc;
 
+    int old_core = to->tcbAffinity;
+    int new_core = sc->scCore;
+
     SMP_COND_STATEMENT(migrateTCB(to, sc->scCore));
+
+    if (old_core != new_core) {
+        SMP_COND_STATEMENT(remoteQueueUpdate(to));
+    }
 }
```
You can see my comments above for the state the thread gets stuck in when changing affinity.

**Comment by Indanz** [comment]
Can this be closed now that b0bdaf9f5 is merged?

**Comment by lsf37** [comment]
I believe it can, yes.
